homeassistant/components/soma/__init__.py

```diff
@@ -120,4 +121,24 @@ async def async_update(self):
             self.is_available = False
             return
         self.current_position = 100 - response["position"]
+        try:
+            response = await self.hass.async_add_executor_job(
+                self.api.get_battery_level, self.device["mac"]
+            )
+        except RequestException:
+            _LOGGER.error("Connection to SOMA Connect failed")
+            self.is_available = False
+            return
+        if response["result"] != "success":
+            _LOGGER.error(
+                "Unable to reach device %s (%s)", self.device["name"], response["msg"]
+            )
+            self.is_available = False
+            return
+        battery = round(2.111 * response["battery_level"] - 763.3)
```

Comment:
What kind of measurement is the original battery value?

Reply:
Found it in the api docs:
> The battery level is in units of 10 mV. The battery level should usually be between 420 and 320. Anything under 320 is critically low. The device itself considers 360 the minimum to move the motor - anything under this will reject all motion commands.

Reply:
What's the logic behind the calculation we do now?

Reply:
I had a number of data points from the Soma Smart Shades iOS app, along with the data from the API, and used that to make the calculation fit.

Smart Shades App percentage | API battery level
----------------------------------|-------------------
98 | 408
72 | 396
60 | 390

Reply:
Seems like this should work. It sets the maximum to 410 and the minimum to 360.
```
Python 3.8.5 (default, Jul 28 2020, 12:59:40) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.18.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: def convert_battery(battery):
   ...:     _battery = (battery - 360) * (100 / 50)
   ...:     return round(min(max(_battery, 0), 100))
   ...: 

In [2]: convert_battery(408)
Out[2]: 96

In [3]: convert_battery(396)
Out[3]: 72

In [4]: convert_battery(390)
Out[4]: 60
```

Reply:
Here's a plot of your measured data. It seems linear so the regular scale mapping formula should work: https://stackoverflow.com/a/11107254

![soma_battery_data](https://user-images.githubusercontent.com/3181692/103172848-bb32fb80-4856-11eb-82df-75ef56feb7be.png)

Reply:
I can change it to that formula if you'd prefer? It gives slightly lower percentages at the top end, but better to have something giving data, and erring on the lower side is probably better for triggering automations to notify that charging is required.

Reply:
Good!